### ISSUE TITLE
Bitmap Subtitles

### ISSUE THREAD

**Issue by lukasf** [lukasf/FFmpegInterop#33]
Today I have been working on adding support for bitmap based subtitles. I worked based on specifications I found on multiple websites. They always explained the idx header format and then said that the actual images are 4-bit palettized bitmaps. But when I ran the code, I found that the bitmap is actually encoded, each sub a different size. And when investigating further, it is a pretty weird kind of encoding. The code alone for decoding sub/idx bitmaps is almost 400 lines of code on a different project. Then I searched for specs of the BluRay format, which I thought was similar to DVD one. But it is again a different format, and also encoded. At least, the BluRay format seems to be a lot easier to decode, but still not as easy as expected.

This website explains the BluRay PGS/SUP format pretty well:
https://github.com/peterdk/SupRip/blob/master/Bluray%20Sup.txt

Here is a (GPL) DVD/VobSub decoder:
https://github.com/mpc-hc/mpc-hc/blob/develop/src/Subtitles/VobSubImage.cpp
("only" lines 1 - 380 are acutally needed for decoding)

I am a bit demotivated now ;)

Still, I just published what I have now, in branch "bitmap-subtitles". I separated the subtitle stuff into a base SubtitleProvider and concrete implementations for Srt, Ssa/Ass, and the (now rather useless) Bitmap implementation. Maybe some of this makes sense to merge into master anyways, even if we do not add Bitmap subtitle support now?

**Comment by mcosmin222** [comment]
Ok. that sounds...daunting.
Let's merge what we have in master first. Even if we do not actually decode the bitmaps yet, it is good to have the separated classes. I was expecting this to be somewhat easier.

**Comment by mcosmin222** [comment]
OK, so I've been playing around and came up with this file.

https://1drv.ms/f/s!Ao-WOwsImxSMtttkizFY7vFeCqivng

According to what I did, this should be a file containing an IDX (or otherwise an image based sub).

The fun part comes now: **ffplay can show the subtitle.** So ffmpeg does know how to do it. If there is an uniform way of handling the images inside ffmpeg, maybe we do not need our own parsers. I suspect these streams should be similar to the ordinary video streams, since there should be no technical difference between a video frame and a subtitle bitmap.

Bonus points: VLC shows the subtitle the same way as ffplay, so they probably use the same parser as well.

I will continue the research on this side.

**Comment by lukasf** [comment]
Oh wow, great find! I searched the ffmpeg docs and found nothing but a subtitle decoder "filter", and that was only available in GPL builts. Looks like avcodec_decode_subtitle2() will do the job! Plus, we can probably also use this to properly decode text based subtitles (so we do not have to parse text and formatting stuff manually). Nice man! :D

**Comment by mcosmin222** [comment]
I am not sure about the text. Depends on the output. I still think it is best to feed text as is because it can scale well with windows DPI and other settings. Maybe for ASS/SSA. But srt or sub should stay text only. We will see once we master this api lol. 

For the moment i haven't had much luck with figuring out how to use decode_subtitle2.

AvSubtitle seems similar to AVFrame, yet i havent found how to init an empty instance. Allocating on stack or with new operator doesn't work properly with decode_subtitle2.

**Comment by mcosmin222** [comment]
Ah yes, I missed the .text field. It will probably be what we are looking for (hopefully). Good point about that.

Any luck with ffplay? I eyeballed it a bit but there doesn't seem to be any corespondent to avsubtitle_aloc. It looks like avcodec_decode_subtitle2 will create the sub if it has a good package.

**Comment by lukasf** [comment]
I had success like this:

```
			AVSubtitle subtitle;
			int gotSubtitle = 0;
			auto result = avcodec_decode_subtitle2(m_pAvCodecCtx, &subtitle, &gotSubtitle, packet);
```

Don't get confused by the result code. A positive value means success, it's the number of bytes used.

**Comment by lukasf** [comment]
It seems that we still need to do some decoding. After debugging quite a while, I think that rect.data[0] will now contain the palettized bitmap and rect.data[1] contains the palette. So we still need to do the palette->rgb converstion. I will see if I can get this running.

**Comment by lukasf** [comment]
Okay I have a first working version ready! But the positioning is still broken (if playback window is resized). We might have to use relative (percentage) positioning. I think we need to know the video size to calculate a proper positioning.

I also hard coded the subtitle image size because I could not find out where it can be read...

**Comment by mcosmin222** [comment]
Nice. good work!

By video size, you mean the size of the video frame sample, or the size of the Media(Player)Element?

**Comment by mcosmin222** [comment]
Can you share the video file on which you tested? It seems the one I created never manages to decode the subtitle packages. At first I thought it was me using the API incorrectly, it probably has some other issue we will handle latter.

**Comment by lukasf** [comment]
I hope we can get correct positioning with only the video frame size (and aspect).

I tested with starwarssub2.vob from this site:
http://samples.mplayerhq.hu/sub/

You need to select the 5th subtitle track.

**Comment by lukasf** [comment]
I already fixed positioning, just set it to 100% w/h instead of pixel size.

Still not sure how to know the acutal size of the subtitle bitmap. We only get the rect size, which usually only covers some part of the image. But how large shall the output bitmap be? I think for dvd/vobsub this is always 720x576. But can we really be sure? It's strage that the size does not seem to be reported anywhere, although it should be part of the subtitle header, so ffmpeg knows about it, theoretically.

It could also be that the file is broken. The samples from that site are usually uploaded to report problems with the player, and this one is from the subtitles category. So maybe it is just that the header is missing in that file, and a correct file would have one.

**Comment by lukasf** [comment]
Pretty weird: If I decode SRT subtitles using avcodec_decode_subtitle2(), then I get a result of type ASS with only .ass filled, .text is null. Looks like ffmpeg will "translate" srt into ass format?! It even creates a default ASS subtitle header, so the resulting ASS could be parsed with the header declaration.

**Comment by mcosmin222** [comment]
What would be the problem if we make the bitmap size the same as the rectangle size? Does it need some sort of "transparent padding" around it?

**Comment by lukasf** [comment]
This does not work. The rects have padding, and usually quite a lot of it, and they would look very bad without it. We know the top/left padding, this is .x and .y of the rect. But we don't know the bottom and right padding. For that we need the total size of the subtitle image. We cannot just guess them or cut off, it would look bad and could cause font size changes and layout changes.

I have just checked ffmpeg sources. It looks like ffmpeg will set width/height from subtitle header, if the header is available. So we only do not see it in this sample because it is a bad file with missing header. Maybe we should use width/height from subtitle AVCodecCtx, and use the video frame size as fallback. This is probably the best we can do. 

I also checked ffmpeg text sub decoders, and indeed all text based subtitle decoders will transcode to ASS in their decode function. So ASS is the common output format of all ffmpeg text subtitle codecs. Maybe that's not such a bad idea (from their point of view), since it is the most powerful format. But it is a bit strange that this cannot be configured, a pure text decode would have been easy as well. At least the format is rather easy to interpret. We could have one class that will handle bitmap subs and one that handles all other subs through decode+ASS parsing.

**Comment by mcosmin222** [comment]
Ok, I see the problem with the bitmap size. I agree that it is probably best to use the AvCodecCtx values with video frame size as fallback. 

Regarding the ASS/SSA issue, it does make sense if they handle this format as a superset of SRT. Now comes the question if we should attempt to continue using our own SRT parser or use theirs. I'd say to go ahead and use their API, at least they probably have a much better time handling all the weird edge cases.

**Comment by lukasf** [comment]
Yes this is what I was proposing in my last post. Let's use ffmpeg conversion and have only one text decoder class based on ASS/SSA format.

**Comment by lukasf** [comment]
Latest commit uses codec props to decide on subtitle provider. This now works with every subtitle format supported by ffmpeg. I tested PGS BluRay format, it works as well. I still left the SRT provider in, because our ASS parsing is pretty poor. We need to parse the header to find out how many commas to skip to get full sub text. Then we can switch all to ASS.

**Comment by mcosmin222** [comment]
Nicely done. I will start testing now...

**Comment by lukasf** [comment]
I have improved the SSA/ASS parser (read header, fixup newline issues) and I think it is now good for the job. At least, I could not find any problems with the files I tried, so I removed the SRT parser and moved everything to SSA/ASS.

I noticed that subtitle font size was different, depending on video resolution. So I changed font size from percentage to fixed pixels, now the size is stable and comparable to what i see in MPC and VLC. I also slightly changed the text outline (some transparency and slightly wider). Let me know what you think.

**Comment by mcosmin222** [comment]
I had a quick look with the files I have, this seems OK. I never noticed the scaling with the video - I mostly use 1080 videos all the time ^^ Nice catch.

Sadly we will have to keep the anti flicker, it seems bitmap subtitles still flicker even with the latest build.

**Comment by lukasf** [comment]
I had some older anime videos where this occured. First I did not even notice, then I tought it might be bitmap subs, but turned out text scaled incorrectly.. ;)

Ok so anti flicker is there to stay. That's unfortunate, but as long as subs don't get dropped, it's probably ok.

**Comment by touseefbsb** [comment]
@lukasf  so the flicker problem is staying in 1803?

**Comment by mcosmin222** [comment]
For now. Maybe we should send feedback on the problem, for vnext?

**Comment by touseefbsb** [comment]
I agree @mcosmin222  this is a crucial issue, and I think a proper uservoice must be created for this

**Comment by lukasf** [comment]
We have the anti-flicker workaround in the lib, so usually subtitles should not flicker. But if we disable the workaround, at least bitmap subtitles will flicker, according to @mcosmin222 .

It would probably be good to report this. But reporting framework bugs to MS is always difficult. They really don't want us to submit bugs. There is no official channel where such bugs could be reported, at least none that I know of.

**Comment by touseefbsb** [comment]
we don't need to submit a framework bug, we can just request it as a feature, maybe they can build the it from scratch and put it in the native sdk. in that case we wont be needing workaround from the framework and we can just use the native functionality. ( for bitmap subs )

**Comment by mcosmin222** [comment]
We can either try through the feedback hub (record to reproduce could be nice to prove that it happens), or in the forum.

I would say the feedback hub is our best bet, I had some very unpleasent interaction with the forum mods apparently being unable to reproduce bugs that everyone else could...

**Comment by touseefbsb** [comment]
@mcosmin222  Do share the link of the post ( of feedback hub ) when u make it, I'll share it as well.

**Comment by mcosmin222** [comment]
I havent encountered any strange (or other) bugs. Looks stable.

**Comment by lukasf** [comment]
Right now I am trying to parse the SSA header's style definitions and create TimedTextRegion and TimedTextStyle based on this. This would allow for custom colors, font style and alignment to work, which are sometimes used on animated stuff. I will add an override switch in the config, so we can still enforce our default style to be used.

**Comment by lukasf** [comment]
Pushed subtitle style parsing. I added support for V4 and V4+ (ASS) styles. Seems to work fine for the files i have here. Output with normal subtitles (SRT) looks more or less same as we had before with our default style. But if real SSA/ASS files are used, colors and styles are applied.

Two things are still missing:

- Use embedded fonts. It should be possible to extract them, put to temp folder and then use for subs.
- Use inline style effects. Subtitle text can contain style info such as {\i1} for italic, etc. This could also be interpreted and used. It will not be possible to do per-word styling, but we could at least make the line italic or bold. Often it is a complete line that is styled. This is not uncommon, I also saw SRT subtitles use html tags for styling lines, and this is translated into inline effects by ffmpeg.

But I will probably save these two for later. Let's fist get this PR finished and out.
